### DISCUSSION TITLE
Button Matrix Helper Example

### DISCUSSION THREAD

**Discussion by brucedk** [tttapa/Control-Surface#526] [Q&A] [answered]
Hello guys, I was making some tests with the library, but I'm new to arduino/C/C++ programing an don't quite understand callback functions on it.

I was testing the Button Matrix Helper [https://tttapa.github.io/Control-Surface-doc/Doxygen/dd/d29/classAH_1_1ButtonMatrix.html#a9d8efbcc279600bd90ce90a76d339856](https://tttapa.github.io/Control-Surface-doc/Doxygen/dd/d29/classAH_1_1ButtonMatrix.html#a9d8efbcc279600bd90ce90a76d339856)

I just want to create a function that when the onButtonChanged callback its called, it prints to serial that change.

my initial code was:

```
#include <AH/Hardware/ButtonMatrix.hpp>

ButtonMatrix<3, 3> btts {
  {6, 7, 8},
  {9, 10, 11}
  };

void setup() {
  // put your setup code here, to run once:
btts.begin();


}

void loop() {
  // put your main code here, to run repeatedly:
btts.update();


}
```
But I don't have a clue where I put the callback...

**Comment by EvgeniySavkin** [comment]
Try using this example:

https://github.com/tttapa/Control-Surface/blob/master/examples/Other/GitHub-issues/Bank-Button-Matrix/Bank-Button-Matrix.ino

**Comment by tttapa** [comment, marked as answer]
The ButtonMatrix class uses a form of static polymorphism known as [CRTP](https://en.wikipedia.org/wiki/Curiously_recurring_template_pattern).

You can use it like this:
```cpp
#include <Arduino_Helpers.h>
#include <AH/Hardware/ButtonMatrix.hpp>

template <uint8_t NumRows, uint8_t NumCols>
class MyButtonMatrix 
    : public AH::ButtonMatrix<MyButtonMatrix<NumRows, NumCols>, NumRows, NumCols> {
  private:
    void onButtonChanged(uint8_t row, uint8_t col, bool  state) {
      Serial << F("Button (") << row << ", " << col << F(") was ") 
             << (state ? F("released") : F("pressed")) << endl;
    }

  public:
    using Parent = typename AH::ButtonMatrix<MyButtonMatrix<NumRows, NumCols>, NumRows, NumCols>;
    using Parent::ButtonMatrix;
    friend Parent;
};

MyButtonMatrix<3, 3> btns {
  {2, 3, 4}, // rows
  {5, 6, 7}, // cols
};

void setup() {
  Serial.begin(115200);
  btns.begin();
}

void loop() {
  btns.update();
}
```
(Note that this example is for the current `master` branch, it won't work with version `1.2.0`, where it run-time polymorphism was used.)

**Reply by brucedk** [reply]
Thanks @tttapa, worked like a charm!

I will now study it better with your example, thanks again
